File: solutions/python/Merge k Sorted Lists.py
class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next
class Solution:
    def merge(self,l1,l2):
        node=ListNode(0)
        temp=node

        while l1 and l2:
            if l1.val<=l2.val:
                temp.next=ListNode(l1.val)
                temp=temp.next
                l1=l1.next
            else:
                temp.next=ListNode(l2.val)
                temp=temp.next
                l2=l2.next
        while l1:
            temp.next=ListNode(l1.val)
            l1=l1.next
            temp=temp.next
        while l2:
            temp.next=ListNode(l2.val)
            l2=l2.next
            temp=temp.next
        return node.next

    def mergeKLists(self, lists: List[Optional[ListNode]]) -> Optional[ListNode]:
        if len(lists)==0:
            return None
        i=0
        last=len(lists)-1

        while last!=0:
            i=0
            j=last
            while i < j:
                lists[i]=self.merge(lists[i],lists[j])
                i=i+1
                j=j-1
                if(i>=j):
                    last=j;
        return lists[0]

File: solutions/cpp/Merge k Sorted Lists.cpp
#include <vector>
using namespace std;

struct ListNode
{
    int val;
    ListNode *next;
    ListNode() : val(0), next(nullptr) {}
    ListNode(int x) : val(x), next(nullptr) {}
    ListNode(int x, ListNode *next) : val(x), next(next) {}
};

class Solution
{
public:
    ListNode *merge(ListNode *a, ListNode *b)
    {
        ListNode *dummy = new ListNode(0);
        ListNode *temp = dummy;
        while (a && b)
        {
            if (a->val < b->val)
            {
                temp->next = new ListNode(a->val);
                temp = temp->next;
                a = a->next;
            }
            else
            {
                temp->next = new ListNode(b->val);
                temp = temp->next;
                b = b->next;
            }
        }

        while (a)
        {
            temp->next = new ListNode(a->val);
            temp = temp->next;
            a = a->next;
        }

        while (b)
        {
            temp->next = new ListNode(b->val);
            temp = temp->next;
            b = b->next;
        }
        return dummy->next;
    };

    ListNode *mergeKLists(vector<ListNode *> &lists)
    {
        if (lists.size() == 0)
            return NULL;
        int last = lists.size() - 1;
        while (last != 0)
        {
            int i = 0;
            int j = last;
            while (i < j)
            {
                lists[i] = merge(lists[i], lists[j]);
                i++;
                j--;
                if (i >= j)
                    last = j;
            }
        }
        return lists[0];
    }
};